### ISSUE TITLE
Tablet table data is not replicated to new node after bootstrap

### ISSUE THREAD

**Issue by asias** [scylladb/scylladb#16706]
Scylla commit: 94580df1c59be7508853e7d99f49177894f1ff86

- Start n1

```
$ nodetool status                                                                                                                                                                                               
Datacenter: dc1                                                                                                                                                                                                                        
===============                                                                                                                                                                                                                        
Status=Up/Down                                                                                                                                                                                                                         
|/ State=Normal/Leaving/Joining/Moving                                                                                                                                                                                                 
--  Address    Load       Tokens       Owns    Host ID                               Rack                                                                                                                                              
UN  127.0.0.1  177.28 KB  256          ?       face9d92-6e43-43e5-8cfd-8eed41918e43  rack1      
```


- Insert data to a tablet table with RF = 2.

```
cqlsh> SELECT * from system.tablets ;

 keyspace_name | table_id                             | last_token           | table_name | tablet_count | new_replicas | replicas                                    | session | stage
---------------+--------------------------------------+----------------------+------------+--------------+--------------+---------------------------------------------+---------+-------
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 | -8070450532247928833 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 | -6917529027641081857 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 | -5764607523034234881 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 | -4611686018427387905 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 | -3458764513820540929 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 | -2305843009213693953 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 | -1152921504606846977 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |                   -1 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  1152921504606846975 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  2305843009213693951 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  3458764513820540927 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  4611686018427387903 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  5764607523034234879 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  6917529027641081855 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  8070450532247928831 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  9223372036854775807 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null

```

- Start n2. Data is not replicated to n2. All tablets have only one replica. 

```
$ nodetool status                                                                                                                                                                                               
Datacenter: dc1                                                                                                                                                                                                                        
===============
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address    Load       Tokens       Owns    Host ID                               Rack
UN  127.0.0.1  177.28 KB  256          ?       face9d92-6e43-43e5-8cfd-8eed41918e43  rack1
UN  127.0.0.2  9.02 MB    256          ?       fe91fcd8-52ad-4cfd-87db-223cd66735a0  rack2

```
```
cqlsh> SELECT * from system.tablets ;

 keyspace_name | table_id                             | last_token           | table_name | tablet_count | new_replicas | replicas                                    | session | stage
---------------+--------------------------------------+----------------------+------------+--------------+--------------+---------------------------------------------+---------+-------
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 | -8070450532247928833 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 | -6917529027641081857 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 | -5764607523034234881 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 | -4611686018427387905 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 | -3458764513820540929 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 | -2305843009213693953 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 | -1152921504606846977 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |                   -1 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  1152921504606846975 |  standard1 |           16 |         null | [(face9d92-6e43-43e5-8cfd-8eed41918e43, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  2305843009213693951 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  3458764513820540927 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  4611686018427387903 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  5764607523034234879 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  6917529027641081855 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  8070450532247928831 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null
           ks1 | 4e47a230-af61-11ee-834e-c2e64f7d7da0 |  9223372036854775807 |  standard1 |           16 |         null | [(fe91fcd8-52ad-4cfd-87db-223cd66735a0, 0)] |    null |  null

(16 rows)
```

**Comment by raphaelsc** [comment]
I can see this happening if

1) created keyspace with initial_tablets 16 and RF=1.
2) created table under that keyspace
3) altered keyspace to RF=2

tablet_allocator_impl, which listens currently only to on_before_create_column_family(), could miss updates to keyspaces as it doesn't override on_update_keyspace(), in order to allocate new replicas if RF has increased.

IIRC, Tomek had a document about supporting changes to RF in tablets.

**Comment by tgrabiec** [comment]
Handling RF changes is a separate issue: https://github.com/scylladb/scylladb/issues/16529

**Comment by raphaelsc** [comment]
The title says after bootstrap, but it seems to me that tablet mapping is created with the 2 nodes alive, but before the RF was changed.

**Comment by raphaelsc** [comment]
Maybe Asias just pasted the wrong tablet map for the scenario before n2 starts. And the map we see is after migration, after n2 started.

So the problem is that tablet allocator doesn't have enough nodes to achieve RF when table is created, and tablet allocator doesn't allocate new tablet replicas after new nodes become available.

@tgrabiec Isn't this all part of RF handling? If we'll know how to allocate or deallocate replicas based on RF changes, it seems like the same infrastructure could be reused to fix this problem, no?

**Comment by tgrabiec** [comment]
> The title says after bootstrap, but it seems to me that tablet mapping is created with the 2 nodes alive, but before the RF was changed.

Asias didn't say anything about changing RF. The tablet map has 1 replica because that's how many nodes were available at the time table was created.

**Comment by tgrabiec** [comment]
> So the problem is that tablet allocator doesn't have enough nodes to achieve RF when table is created, and tablet allocator doesn't allocate new tablet replicas after new nodes become available.

The problem here is that replicas are not moved to the new node.

**Comment by asias** [comment]
> @asias Do you have logs from n1?

It is super easy to reproduce. It is reproduced 100%.  Just two nodes.

**Comment by asias** [comment]
As mentioned exactly in the opening message, there is no changing of RF at all.  Start n1. Insert data with RF = 2 (At this time there is only one node in the cluster). Add n2. I would expect scylla replicate data to n2. There are two level of problems, 1) the data ownership is not changed after n2 is added. 2) Data is not streamed to n2 after data ownership change.

**Comment by xemul** [comment]
FYI -- this setup is going to be rejected after #17079

**Comment by avikivity** [comment]
Yes, that's better than #16529 as it's clear what the cluster guarantees wrt replication and quorum.

**Comment by denesb** [comment]
@xemul I assigned you because you are already working on a fix (https://github.com/scylladb/scylladb/pull/17079).

**Comment by tgrabiec** [comment]
> @xemul I assigned you because you are already working on a fix (#17079).

That's a different problem (https://github.com/scylladb/scylladb/issues/16529). The problem here is that replicas stay on the old host, while it should be balanced between nodes.

**Comment by tgrabiec** [comment]
> > @xemul I assigned you because you are already working on a fix (#17079).
> 
> That's a different problem (#16529). The problem here is that replicas stay on the old host, while they should be balanced between nodes.

Actually, looking at the output of select from system.tablets, there are two replicas used: fe91fcd8-52ad-4cfd-87db-223cd66735a0 and face9d92-6e43-43e5-8cfd-8eed41918e43. So there is no other issue.

**Comment by tgrabiec** [comment]
Duplicate of https://github.com/scylladb/scylladb/issues/16529

**Comment by tgrabiec** [comment]
@asias The output of system.tablets after step `Insert data to a tablet table with RF = 2.` must not be the right one, since it already contains replica on n2, which is supposedly added later.
